zuul-core/src/main/java/com/netflix/zuul/message/http/HttpRequestMessageImpl.java

```diff
@@ -510,7 +511,10 @@ protected String generateInfoForLogging() {
     @Override
     public String getOriginalHost() {
         try {
-            return getOriginalHost(getHeaders(), getServerName());
+            if (originalHost == null) {
+                originalHost = getOriginalHost(getHeaders(), getServerName());
```

Comment:
The compute intensive part here should be the `parseHostHeader` logic. 
Consider the option to just cache that part and not the `X-Forwarded-Host` part.

Reply:
`X-Forwarded-Host` part is called everytime before the `parseHostHeader` and it loops on all existing headers to check for the `x-forwarded-host` using `headers.getFirst` before attempting to call the `parseHostHeader`. Doesn't sound efficient. I am tempted to cache that as well.

Reply:
hmm, it does iterate through the list, ugh.